src/transport/SecureSessionMgr.cpp

```diff
@@ -92,136 +92,157 @@ CHIP_ERROR SecureSessionMgr::SendMessage(NodeId peerNodeId, System::PacketBuffer
+
+    msgBuf->SetStart(msgBuf->Start() - headerSize);
+    data     = msgBuf->Start();
+    totalLen = msgBuf->TotalLength();
+
+    err = payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize);
+    SuccessOrExit(err);
 
-    if (isEncrypted)
-    {
-        err = packetHeader.Decode(msgBuf->Start(), msgBuf->DataLength(), &headerSize);
-        SuccessOrExit(err);
+    err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
+    SuccessOrExit(err);
+    err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
+    SuccessOrExit(err);
+
+    VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
+    msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen));
 
-        VerifyOrExit(packetHeader.GetDestinationNodeId().HasValue(), err = CHIP_ERROR_INVALID_DESTINATION_NODE_ID);
-        peerNodeId = packetHeader.GetDestinationNodeId().Value();
+    ChipLogDetail(Inet, "Secure transport encrypted msg %u", msgId);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        const char * errStr = ErrorStr(err);
+        ChipLogError(Inet, "Secure transport failed to encrypt msg %u: %s", state->GetSendMessageIndex(), errStr);
+    }
+    else
+    {
+        state->IncrementSendMessageIndex();
     }
 
-    state = mPeerConnections.FindPeerConnectionState(peerNodeId, nullptr);
+    return err;
+}
+
+CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, PacketHeader & packetHeader, NodeId peerNodeId,
+                                         System::PacketBufferHandle msgBuf, EncryptedPacketBufferHandle * bufferRetainSlot,
+                                         bool isEncrypted)
+{
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    PeerConnectionState * state = nullptr;
+    uint8_t * msgStart          = nullptr;
+    uint16_t msgLen             = 0;
+    uint16_t headerSize         = 0;
+
+    // Hold the reference to encrypted message in stack variable.
+    // In case of any failures, the reference is not returned, and this stack variable
+    // will automatically free the reference on returning from the function.
+    EncryptedPacketBufferHandle encryptedMsg;
+
     VerifyOrExit(mState == State::kInitialized, err = CHIP_ERROR_INCORRECT_STATE);
 
     VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrExit(!msgBuf->HasChainedBuffer(), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
     VerifyOrExit(msgBuf->TotalLength() < kMax_SecureSDU_Length, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
 
     // Find an active connection to the specified peer node
+    state = mPeerConnections.FindPeerConnectionState(peerNodeId, nullptr);
     VerifyOrExit(state != nullptr, err = CHIP_ERROR_INVALID_DESTINATION_NODE_ID);
 
     // This marks any connection where we send data to as 'active'
     mPeerConnections.MarkConnectionActive(state);
 
+    if (!isEncrypted)
```

Comment:
nit: I would find it a bit easier to read the positive `if (encryptPayload)` instead of the negated `if (!isEncrypted)`.